File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        String res = "";
        int l = 0, r = 0;
        for(int i = 0; i < s.length() - 1; i++) {
            int len1 = panlindrome(s, i, i);
            int len2 = panlindrome(s, i, i + 1);
            int len = Math.max(len1, len2);
            if(len > r - l) {
                //for odd length, (len -1) /2 is the left mid point of panlindrome, len /2 is right mid point
                //for even length, len/2 and (len - 1) /2 is same mid point
                l = i - (len - 1) / 2;
                r = i + len/ 2;
            }
        }
        return s.substring(l, r + 1);
    }
    public int panlindrome(String s, int l , int r) {
        while(l >= 0 && r < s.length()) {
            if(s.charAt(l) == s.charAt(r)) {
                l--;
                r++;
            } else break;
        }
        return r - l - 1;
    }
}

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        #中心扩散
        res = 0
        left, right = 0, 0
        for i in range(len(s)):
            def find(s, l, r):
                while(l >= 0 and r < len(s) and s[l] == s[r]):
                    l -= 1
                    r += 1
                return r - l - 1
            ol = find(s, i, i)
            el = find(s, i, i + 1)
            ml = max(ol, el)
            if ml  > res:
                res = ml
                left =  i - (ml - 1) // 2
                right = i + (ml) // 2
                
        return s[left: right + 1]